qiskit/visualization/pulse_v2/events.py

```diff
@@ -0,0 +1,204 @@
+than plotting each operand value bound to the instruction.
+"""
+from collections import defaultdict, namedtuple
+from typing import Dict, List, Iterator, Tuple
+
+from qiskit import pulse
+
+PhaseFreqTuple = namedtuple('PhaseFreqTuple', 'phase freq')
+
+
+class ChannelEvents:
+    """Channel event manager.
+    """
+    _waveform_group = pulse.Play, pulse.Delay, pulse.Acquire
+    _frame_group = pulse.SetFrequency, pulse.ShiftFrequency, pulse.SetPhase, pulse.ShiftPhase
+    _misc_group = pulse.Snapshot
+
+    def __init__(self,
+                 waveforms: Dict[int, pulse.Instruction],
+                 frames: Dict[int, List[pulse.Instruction]],
+                 miscellaneous: Dict[int, List[pulse.Instruction]],
+                 channel: pulse.channels.Channel):
+        """Create new event manager.
+
+        Args:
+            waveforms: List of waveforms shown in this channel.
+            frames: List of frame change type instructions shown in this channel.
+            miscellaneous: List of miscellaneous instructions shown in this channel.
+            channel: Channel object associated with this manager.
+        """
+        self._waveforms = waveforms
+        self._frames = frames
+        self._miscellaneous = miscellaneous
+        self.channel = channel
+
+        # initial frame
+        self.init_phase = 0
+        self.init_frequency = 0
+
+    @classmethod
+    def parse_program(cls,
+                      program: pulse.Schedule,
+                      channel: pulse.channels.Channel):
+        """Parse pulse program represented by ``Schedule``.
+
+        Args:
+            program: Target ``Schedule`` to visualize.
+            channel: The channel managed by this instance.
+        """
+        waveforms = dict()
+        frames = defaultdict(list)
+        miscellaneous = defaultdict(list)
+
+        # parse instructions
+        for t0, inst in program.filter(channels=[channel]).instructions:
+            if isinstance(inst, cls._waveform_group):
+                waveforms[t0] = inst
+            elif isinstance(inst, cls._frame_group):
+                frames[t0].append(inst)
+            elif isinstance(inst, cls._misc_group):
+                miscellaneous[t0].append(inst)
+
+        return ChannelEvents(waveforms, frames, miscellaneous, channel)
+
+    def is_empty(self):
+        """Check if there is any nonzero waveforms in this channel.
+        """
+        for waveform in self._waveforms.values():
+            if isinstance(waveform, (pulse.Play, pulse.Acquire)):
+                return False
+        else:
+            return True
+
+    def get_waveforms(self) -> Iterator[Tuple[int, PhaseFreqTuple, pulse.Instruction]]:
+        """Return waveform type instructions with frame.
+        """
+        sorted_frame_changes = sorted(self._frames.items(), key=lambda x: x[0], reverse=True)
+
+        # bind phase and frequency with instruction
+        phase = self.init_phase
+        frequency = self.init_frequency
```

Comment:
setting frequency requires backend object. so I avoid to add it to to this layer. because this class is called only by the other program, frequency info can be directly set by interface function. I assume the initial phase is usually zero.